File: dotnet/src/MableBank.Domain/TransferInstruction.cs
using MableBank.Domain.Errors;

namespace MableBank.Domain;

public sealed record TransferInstruction(
    TransferId TransferId,
    AccountNumber FromAccountNumber,
    AccountNumber ToAccountNumber,
    Money Amount)
{
    public static TransferInstruction Create(
        AccountNumber fromAccountNumber,
        AccountNumber toAccountNumber,
        Money amount,
        TransferId? transferId = null) =>
        CreateWithId(transferId ?? TransferId.Derive(fromAccountNumber, toAccountNumber, amount), fromAccountNumber, toAccountNumber, amount);

    public static TransferInstruction CreateWithId(
        TransferId transferId,
        AccountNumber fromAccountNumber,
        AccountNumber toAccountNumber,
        Money amount)
    {
        if (amount.Amount == 0m)
        {
            throw new InvalidTransferInstructionException("Transfer amount must be positive");
        }

        if (fromAccountNumber.Equals(toAccountNumber))
        {
            throw new InvalidTransferInstructionException("Source and destination accounts must differ");
        }

        return new TransferInstruction(transferId, fromAccountNumber, toAccountNumber, amount);
    }
}

File: nodejs-typescript/src/domain/TransferInstruction.ts
import type { AccountNumber } from './AccountNumber.js';
import { InvalidTransferInstructionError } from './errors.js';
import type { Money } from './Money.js';
import { TransferId } from './TransferId.js';

export type TransferInstruction = {
  readonly transferId: TransferId;
  readonly fromAccountNumber: AccountNumber;
  readonly toAccountNumber: AccountNumber;
  readonly amount: Money;
};

export function createTransferInstruction(input: {
  fromAccountNumber: AccountNumber;
  toAccountNumber: AccountNumber;
  amount: Money;
  transferId?: TransferId;
}): TransferInstruction {
  if (input.amount.isZero()) {
    throw new InvalidTransferInstructionError('Transfer amount must be positive');
  }
  if (input.fromAccountNumber.equals(input.toAccountNumber)) {
    throw new InvalidTransferInstructionError('Source and destination accounts must differ');
  }
  const transferId =
    input.transferId ??
    TransferId.derive(input.fromAccountNumber, input.toAccountNumber, input.amount);
  return {
    transferId,
    fromAccountNumber: input.fromAccountNumber,
    toAccountNumber: input.toAccountNumber,
    amount: input.amount,
  };
}
